File: solutions/python/resistor-color-trio/resistor_color_trio.py
from typing import Sequence

COLOR_MAP = {
    'black': 0, 'brown': 1, 'red': 2, 'orange': 3, 'yellow': 4,
    'green': 5, 'blue': 6, 'violet': 7, 'grey': 8, 'white': 9
}

SUFFIXES = (
    (1_000_000_000, 'gigaohms'),
    (1_000_000, 'megaohms'),
    (1_000, 'kiloohms'),
)

def label(colors: Sequence[str]) -> str:
    significant_digits = COLOR_MAP[colors[0]] * 10 + COLOR_MAP[colors[1]]
    multiplier = 10 ** COLOR_MAP[colors[2]]
    ohms = significant_digits * multiplier

    for divisor, suffix in SUFFIXES:
        if ohms >= divisor:
            return f'{ohms // divisor} {suffix}'

    return f'{ohms} ohms'

File: solutions/java/resistor-color-trio/src/main/java/ResistorColorTrio.java
import java.util.Arrays;
import java.util.List;

class ResistorColorTrio {
    private static final List<String> COLORS = List.of(
            "black", "brown", "red", "orange", "yellow",
            "green", "blue", "violet", "grey", "white"
    );

    String label(String[] colors) {
        var colorValues = Arrays.stream(colors)
                .mapToInt(COLORS::indexOf)
                .toArray();
        var value = colorValues[0] * 10 + colorValues[1];
        var power = (long) Math.pow(10, colorValues[2]);
        var resistance = value * power;

        String unit = "ohms";
        if (resistance >= 1_000_000_000) {
            unit = "gigaohms";
            resistance /= 1_000_000_000;
        } else if (resistance >= 1_000_000) {
            unit = "megaohms";
            resistance /= 1_000_000;
        } else if (resistance >= 1000) {
            resistance /= 1000;
            unit = "kiloohms";
        }
        return resistance + " " + unit;
    }
}
